vm-virtio/src/console.rs

```diff
@@ -0,0 +1,389 @@
+}
+
+/// Input device.
+pub struct ConsoleInput {
+    input_evt: EventFd,
+    in_buffer: Arc<Mutex<VecDeque<u8>>>,
+}
+
+impl ConsoleInput {
+    pub fn queue_input_bytes(&self, input: &[u8]) {
+        let mut in_buffer = self.in_buffer.lock().unwrap();
+        in_buffer.extend(input);
+        let _ = self.input_evt.write(1);
+    }
+}
+
+impl Console {
+    /// Create a new virtio console device that gets random data from /dev/urandom.
+    pub fn new(out: Option<Box<io::Write + Send>>) -> io::Result<(Console, Arc<ConsoleInput>)> {
+        let avail_features = 1u64 << VIRTIO_F_VERSION_1;
+
+        let input_evt = EventFd::new(EFD_NONBLOCK).unwrap();
+
+        let console_input = Arc::new(ConsoleInput {
+            input_evt,
+            in_buffer: Arc::new(Mutex::new(VecDeque::new())),
+        });
+
+        Ok((
+            Console {
+                kill_evt: None,
+                avail_features,
+                acked_features: 0u64,
+                input: console_input.clone(),
+                out,
+            },
+            console_input,
+        ))
+    }
+}
+
+impl Drop for Console {
+    fn drop(&mut self) {
+        if let Some(kill_evt) = self.kill_evt.take() {
+            // Ignore the result because there is nothing we can do about it.
+            let _ = kill_evt.write(1);
+        }
+    }
+}
+
+impl VirtioDevice for Console {
+    fn device_type(&self) -> u32 {
+        VirtioDeviceType::TYPE_CONSOLE as u32
+    }
+
+    fn queue_max_sizes(&self) -> &[u16] {
+        QUEUE_SIZES
+    }
+
+    fn features(&self, page: u32) -> u32 {
+        match page {
+            // Get the lower 32-bits of the features bitfield.
+            0 => self.avail_features as u32,
+            // Get the upper 32-bits of the features bitfield.
+            1 => (self.avail_features >> 32) as u32,
+            _ => {
+                warn!("Received request for unknown features page.");
+                0u32
+            }
+        }
+    }
+
+    fn ack_features(&mut self, page: u32, value: u32) {
+        let mut v = match page {
+            0 => u64::from(value),
+            1 => u64::from(value) << 32,
+            _ => {
+                warn!("Cannot acknowledge unknown features page.");
+                0u64
+            }
+        };
+
+        // Check if the guest is ACK'ing a feature that we didn't claim to have.
+        let unrequested_features = v & !self.avail_features;
+        if unrequested_features != 0 {
+            warn!("Received acknowledge request for unknown feature.");
+
+            // Don't count these features as acked.
+            v &= !unrequested_features;
+        }
+        self.acked_features |= v;
+    }
+
+    fn read_config(&self, _offset: u64, _data: &mut [u8]) {
+        warn!("No currently device specific configration defined");
+    }
+
+    fn write_config(&mut self, _offset: u64, _data: &[u8]) {
+        warn!("No currently device specific configration defined");
```

Comment:
I don't understand this message (and it has a typo.)

Reply:
We don't support the device configuration structure yet. We will have it in the next iteration. I have changed the comment.